### ISSUE TITLE
Field of union needs to be different type when reading vs setting

### ISSUE THREAD

**Issue by Zeckie** [python/mypy#16732]
**Bug Report**

Mypy appears to infer that the type of a field of a variable that is typed as a union is a single type whether that field is being read or updated. The type correct when reading the field (union of the types of the fields) , but not correct when setting as it does not enforce that the value being assigned needs to be valid for both of the types in the union (intersection of the types of the fields).

**To Reproduce**

```python
class Foo:
    a: int

class Bar:
    a: str

either: Foo | Bar
either.a = 1  # No error, despite not being valid for Bar
```
More detailed example: https://mypy-play.net/?mypy=latest&python=3.12&gist=bf74f1ba35979f10b1878b2e1abca865

**Expected Behavior**

Mypy should report the assignment as an error if it does not work with all the types in the union.

**Actual Behavior**

No error, despite the assignment allowing a field of type `str` to be an `int`
